drivers/char/rust_example.rs

```diff
@@ -51,23 +53,120 @@ module! {
     },
 }
 
-struct RustFile;
+const MAX_TOKENS: usize = 3;
 
-impl FileOperations for RustFile {
+struct SharedStateInner {
+    token_count: usize,
+}
+
+struct SharedState {
+    state_changed: CondVar,
+    inner: Mutex<SharedStateInner>,
+}
+
+impl SharedState {
+    fn try_new() -> KernelResult<Arc<Self>> {
+        let state = Arc::try_new(Self {
+            // SAFETY: `condvar_init!` is called below.
+            state_changed: unsafe { CondVar::new() },
+            // SAFETY: `mutex_init!` is called below.
+            inner: unsafe { Mutex::new(SharedStateInner { token_count: 0 }) },
+        })?;
+        // SAFETY: `state_changed` is pinned behind `Arc`.
+        let state_changed = unsafe { Pin::new_unchecked(&state.state_changed) };
+        kernel::condvar_init!(state_changed, "SharedState::state_changed");
+        // SAFETY: `inner` is pinned behind `Arc`.
```

Comment:
This function needs to return `KernelResult<Pin<Arc<Self>>>` for `state_changed` and `inner` to be pinned. Otherwise you could take two `Arc<Self>` and then use `mem::swap(Arc::get_mut(a).unwrap(), Arc::get_mut(b).unwrap())` to swap both fields between two allocations.